src/utils/src/Time.c

```diff
@@ -3,12 +3,76 @@
 getTime globalGetTime = defaultGetTime;
 getTime globalGetRealTime = defaultGetTime;
 
+STATUS generateTimestampStrInMilliseconds(PCHAR formatStr, PCHAR pDestBuffer, UINT32 destBufferLen, PUINT32 pFormattedStrLen)
+{
+    STATUS retStatus = STATUS_SUCCESS;
+    UINT32 formattedStrLen;
+    UINT32 millisecondVal;
+    struct tm* timeinfo;
+
+    CHK(pDestBuffer != NULL && pFormattedStrLen != NULL, STATUS_NULL_ARG);
+    CHK(STRNLEN(formatStr, MAX_TIMESTAMP_FORMAT_STR_LEN + 1) <= MAX_TIMESTAMP_FORMAT_STR_LEN, STATUS_MAX_TIMESTAMP_FORMAT_STR_LEN_EXCEEDED);
+
+#if defined _WIN32 || defined _WIN64
+    FILETIME fileTime;
+    GetSystemTimeAsFileTime(&fileTime);
+
+    // Convert the file time to 100-nanosecond intervals (since January 1, 1601)
+    ULARGE_INTEGER uli;
+    uli.LowPart = fileTime.dwLowDateTime;
+    uli.HighPart = fileTime.dwHighDateTime;
+    ULONGLONG fileTime100ns = uli.QuadPart;
+    fileTime100ns -= TIME_DIFF_UNIX_WINDOWS_TIME;
+
+    // Convert to milliseconds
+    ULONGLONG milliseconds100ns = fileTime100ns / 10000;
+
+    // Convert milliseconds to time_t
+    time_t seconds = milliseconds100ns / 1000;
+
+    // Convert time_t to UTC tm struct
+    timeinfo = GMTIME(&seconds);
+
+    millisecondVal = milliseconds100ns % 1000;
+
+#elif defined __MACH__ || defined __CYGWIN__
+    struct timeval tv;
+    gettimeofday(&tv, NULL);  // Get current time
+
+    timeinfo = GMTIME(&(tv.tv_sec));  // Convert to broken-down time
+
+    millisecondVal = tv.tv_usec / HUNDREDS_OF_NANOS_IN_A_MICROSECOND;  // Convert microseconds to milliseconds
+
+#else
+    struct timespec nowTime;
+    clock_gettime(CLOCK_REALTIME, &nowTime);
+
+    timeinfo = GMTIME(&(nowTime.tv_sec));
+    millisecondVal = nowTime.tv_nsec / DEFAULT_TIME_UNIT_IN_NANOS;  // Convert nanoseconds to milliseconds
+
+#endif
+    formattedStrLen = 0;
+    *pFormattedStrLen = 0;
+
+    formattedStrLen = (UINT32) STRFTIME(pDestBuffer, destBufferLen - MAX_MILLISECOND_PORTION_LENGTH, formatStr, timeinfo);
+    CHK(formattedStrLen != 0, STATUS_STRFTIME_FALIED);
+    // Total length is 8 plus terminating null character. Generated string would have utmost size - 1. Hence need to add 1 to max length
+    SNPRINTF(pDestBuffer + formattedStrLen, MAX_MILLISECOND_PORTION_LENGTH + 1, ".%06d ", millisecondVal);
```

Comment:
I see ok so no matter what format str is passed in you will add a `.ssssss` to it.  

Can you make that clear in the API docs or comments, because this function is actually appending things to the user passed in format string, where I would have expected if the user can specify a string they would be required to specify the ms component as well and we would format our output to match what the user specified.  I would prefer we didn't append to user provided format string but I think it's fine in this case as long as it's clear to the user of this function that is what is going to happen to their provided string.  I can see why this is the case since the `tm` struct does not seem to contain ms data so I suppose you cannot use it with `strftime`.

Reply:
Also since this is a public API we need to make sure it works for *any* format str the customer passes and if it doesn't to make sure we state what isn't support.  Since you're appending the ".%06d" to the end, for example if the customer only wanted minute precision and specified that as the format string, i.e. `"%Y-%m-%d %H:%M"` this this will not work you'll get a time like:  "2023-07-11 16:26.123456" in which case the meaning is `.123456` of a minute but the value printed is the millisecond so this will be incorrect.